docs/spec.md

```diff
@@ -160,10 +163,50 @@ following constraints:
 * (C12) If `quantization_dimension` is not empty, then
   `0 <= quantization_dimension`.
 
-At the moment, `QuantizationScale` is a floating-point constant, but there is
-strong interest in integer-based scales, represented with multipliers and
-shifts. We are planning to explore this in the near future
-([#1404](https://github.com/openxla/stablehlo/issues/1404)).
+In order to allow operation using only integer arithmetic, the floating-point
+`scale`, `S`, is realized using integer `multipler`, `M` and `shift` value,
```

Comment:
Either (1) need to explicitly specify `int32_t` multiplier/shift or (2) need to update the syntax to allow different multiplier/shift type (and update the C++ code below).

Reply:
Hi @loganchien 

I have updated the above line to allow multiplier/shift with generic bit widths to provide a generic description of the conversion and the relationship between them. In the C++ code below, I have demonstrated the scale to (multiplier, shift) conversions using a specific bit width.

If we decide on making the code generic, then we might have to decide on a common type to store the values for the multiplier and shift. For example, both [computeMultiplierAndShiftTosaScale32](https://github.com/llvm/llvm-project/blob/94058c41d43a4aae0e0914b151c166f2729affbe/mlir/lib/Dialect/Tosa/Utils/QuantUtils.cpp#L58) and [computeMultiplierAndShiftTosaScale16](https://github.com/llvm/llvm-project/blob/94058c41d43a4aae0e0914b151c166f2729affbe/mlir/lib/Dialect/Tosa/Utils/QuantUtils.cpp#L22) uses `int32_t`  to store those values. For scaling other than 32-bit and 16-bits, it will be more of an implementation details. Please let me know your opinion.

One general question: Is 32-bit and 16-bit scaling (as show in the links above) the only ones supported while legalizing to TOSA? Or it could be completely arbitrary. 
cc @dominicsymes

Reply:
Thanks for the updates.

TOSA rescale supports 32-bit and 16-bit multiplier as you say. It seems unlikely to me that multipliers larger than 32-bit would be used since quantization is usually done to reduce data type size. So, 32-bit I would think is an upper bound. Multipliers smaller than 16-bit may be useful to consider. These can be legalized to the 16-bit case with a smaller value of the multiplier.

In terms of legalization to TOSA and comment from @loganchien below on operations, I think many operations will not use these multiplier and shift values directly, but a derived multiplier and shift specific to that operation. For example, for dot product the output scale will be based on (input1_scale * input2_scale / output_scale). This derivation may want to reduce the number of bits used, for example reducing from 32 to 16 bits.

In terms of example usage for an operation, such as dot product discussed around question 2. of the original comment, can this be considered here, or do you think this will be another PR?

Reply:
Thanks @dominicsymes for explaining. 

>> In terms of example usage for an operation, such as dot product discussed around question 2. of the original comment, can this be considered here, or do you think this will be another PR?

Yes, I am planning to consider it here in the current PR. I am currently looking into the details and planning to post the update asap.

Reply:
@dominicsymes 

Regarding your comment on: `I think it would be useful to give an example, perhaps based around the dot product operation of https://github.com/openxla/stablehlo/pull/1413 since in this operation only a single scale is needed of the final accumulator.`

I was considering a few options here. One is:

Given a dot.general op with floating-point scale as
```
%result = "stablehlo.dot_general"(%lhs, %rhs) {
  dot_dimension_numbers = #stablehlo.dot<
    lhs_batching_dimensions = [0],
    rhs_batching_dimensions = [0],
    lhs_contracting_dimensions = [2],
    rhs_contracting_dimensions = [1]
  >,
  precision_config = [#stablehlo<precision DEFAULT>, #stablehlo<precision DEFAULT>]
} : (tensor<2x2x2x!quant.uniform<i8<-127:127>:f32, 1.25:4>>,
     tensor<2x2x2x!quant.uniform<i8<-127:127>:f32, 1.25:4>>) ->
     tensor<2x2x2x!quant.uniform<i8<-127:127>:f32, 1.25:4>>
```

Applying the scale to 32-bit (multiplier,shift) conversion, as mentioned above, we have

```
%result = "stablehlo.dot_general"(%lhs, %rhs) {
  dot_dimension_numbers = #stablehlo.dot<
    lhs_batching_dimensions = [0],
    rhs_batching_dimensions = [0],
    lhs_contracting_dimensions = [2],
    rhs_contracting_dimensions = [1]
  >,
  precision_config = [#stablehlo<precision DEFAULT>, #stablehlo<precision DEFAULT>]
} : (tensor<2x2x2x!quant.uniform<i8<-127:127>:f32, <1342177280, 30>:4>>, 
     tensor<2x2x2x!quant.uniform<i8<-127:127>:f32, <1342177280, 30>:4>>) ->
     tensor<2x2x2x!quant.uniform<i8<-127:127>:f32, <1342177280, 30>:4>>
```

My intention was to show how the above conversion to this new type format can be used to enable all integer computation. I am not sure if rescale operations help here to capture that.  

Another options is to refer to the  semantic of `dot_general` op (quoted below) 

```
For is_quantized_tensor(lhs) and is_quantized_tensor(rhs):
  - integer_dot_product = reduce( inputs=[multiply((reshaped_lhs_slice - zero_point(reshaped_lhs_slice)), (reshaped_rhs_slice - zero_point(reshaped_rhs_slice))], init_values=[0], dimensions=[0, ..., size(lhs_contracting_dimensions) - 1], body=lambda x, y: add(x, y)).
  - rounded_dot_product = round_nearest_even(integer_dot_product * (scale(reshaped_lhs_slice) * scale(reshape_rhs_slice) / scale(result))).   
   - dot_product = clamp(storage_min(result), rounded_dot_product + zero_point(result), storage_max(result)).
```

and say that the expression
```
scale(reshaped_lhs_slice) * scale(reshape_rhs_slice) / scale(result)
```

can be expressed as the multiplier and shit. But, IMO, this does not demonstrate the benefit of this newly introduced  type. 

I think what I am presenting follow your comment " I think many operations will not use these multiplier and shift values directly, but a derived multiplier and shift specific to that operation. For example, for dot product the output scale will be based on (input1_scale * input2_scale / output_scale)."

Please let me know if I am missing something.

Reply:
Thanks @dominicsymes, a lot, for the explanation. This clarifies a lot of my dilemma. 

One thing to double check:
>> (numerator_M/denominator_M) * 2^(-total_n) = M * 2^(-n)

The operation `(numerator_M/denominator_M)`  should use integer arithmetic, right?  IMO, that would enable an all integer computation.

Reply:
Hi @dominicsymes 
  
I have attached the revised write-up. PTAL.
I have also added some clarity on a few things:

1. Can we have a mix of quantized types with floating-point and integer scales? I think the answer is No.

2. Given, the example we provided for `dot_general` (in the updated PR), where we have shown how the "derived floating-point scale" (which is `(scale(reshaped_lhs_slice) * scale(reshape_rhs_slice) / scale(result))`) is converted to "derived integer scale", one might be tempted to think that, for [addOp](https://github.com/openxla/stablehlo/blob/main/docs/spec.md#add), the derived floating-point scale is

```
scale(lhs)/scale(result) for re-scaling lhs, and
scale(rhs)/scale(result) for re-scaling lhs.
```

which may not match what is done in [practice](https://groups.google.com/a/openxla.org/g/openxla-discuss/c/iwE9is49SS4/m/VTBpjwDrBAAJ) for TOSA re-scaling.  I remember that we already discussed about this,  [here](https://github.com/openxla/stablehlo/pull/1413#discussion_r1185585993), and I thought about clarifying it in the spec as well.

Reply:
Hi @sdasgup3 

> The operation (numerator_M/denominator_M) should use integer arithmetic, right? IMO, that would enable an all integer computation.

Yes, it should be an integer divide but I think there are some issues involving overflow/underflow which I’ve commented on in the new patch.

For (1) do you mean within an operation or within a graph? Within an operation this seems unnecessary and would complicate. Within a graph it may be useful provided there are operations to convert between the two types.

For (2) yes, I think it is worth to clarify this as well for addOp. Scaling directly to result scale for add can lead to precision loss. For example, if output scale is 1.0 and input scales are 0.4 then doing sum of 0.4+0.4 at input scale will give 0.4 + 0.4 = 0.8 round to 1.0 at output scale. Doing sum at output scale would give 0 + 0 = 0. The [TFLite legalization](https://github.com/tensorflow/tensorflow/blob/60f7a770d64cb6cb3a93f84c291272ec51304d31/tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc#L695-L700) discussed in the thread you point to chooses an add scale which does not overflow the integer add but also has extra precision. For example, the TFLite legalization does the add at scale 2*max(scale(lhs), scale(rhs))/(1<<n) where n=20 for 8-bit quantization and n=15 for 16-bit quantization. For these scales the add will not overflow 32 bits.

Reply:
+1 on clarifying the scale issue in the add op

Reply:
Thanks @dominicsymes! 

>> Yes, it should be an integer divide but I think there are some issues involving overflow/underflow which I’ve commented on in the new patch.

Yes, I agree. Looking into it. Thanks!

>> For (1) do you mean within an operation or within a graph? Within an operation this seems unnecessary and would complicate. Within a graph it may be useful provided there are operations to convert between the two types.

I agree with you. 
Actually I meant at both the levels. At first I thought that mixing float vs int scales is equally bad at any level, but realized that in the current form we are specifying the semantics of an op when all the scales, of operands and outputs,  are the same (example add or dot_general ops). With that we cannot  afford having the mix within an operation, without adding addition complexity in specification (for example, what is the semantics of add with lhs and rhs scales are of different "types"). However, at program (graph) level the mixing seems OK.

>> For (2)

I agree with your point about the precision loss. 

Currently, I am trying to answer a few open questions related to the following (also open for discussion ) and get back asap. 
 
1. To what extend we should do the exemplification. All the ops (involving some numeric computation) ? 
2. Can we specify the  "derived scale" in a generic manner, without getting tied to a specific implementation (like tfl legalization to TOSA). Specifically in cases, like AddOp,when the derived integer scale is not intuitive from the derived floating-point scale.